### ISSUE TITLE
SDNExpress - FrontEndIP parameter for gateways is ignored

### ISSUE THREAD

**Issue by MassimoPascucci** [microsoft/SDN#556]
Using SDNExpress with a configuration file (the GUI has no options for static IP addresses).

If you use the FrontEndIP parameter to statically configure the front-end IP address for gateways, the parameter is ignored and automatic assignment from the PA subnet address pool is performed.

Sample config:

    ManagementSubnet     = '10.10.10.0/24'

    PASubnet             = '10.20.30.0/24'
    PAPoolStart          = '10.20.30.100'
    PAPoolEnd            = '10.20.30.200'  

    Gateways = @(
       @{ComputerName='gw01'; HostName='host01'; ManagementIP='10.10.10.131'; FrontEndIP='10.20.30.131'},
       @{ComputerName='gw02'; HostName='host02'; ManagementIP='10.10.10.132'; FrontEndIP='10.20.30.132'}
    )
    
Expected behaviour:
- gw01 should use front-end IP address 10.20.30.131.
- gw02 should use front-end IP address 10.20.30.132.

Actual behaviour:
- Each gateway gets the first available IP address from the PA pool, starting from the end.

**Comment by AnirbanPaul** [comment]
Yes, that is right. We have changed the logic to pick this up automatically from the back of the pool to avoid conflicts with other PA reservations.

Do you have any strong reason to allocate a fixed IP Address from the pool?

Thanks
Anirban

From: Massimo Pascucci ***@***.***>
Sent: Wednesday, January 18, 2023 10:07 AM
To: microsoft/SDN ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [microsoft/SDN] SDNExpress - FrontEndIP parameter for gateways is ignored (Issue #556)


Using SDNExpress with a configuration file (the GUI has no options for static IP addresses).

If you use the FrontEndIP parameter to statically configure the front-end IP address for gateways, the parameter is ignored and automatic assignment from the PA subnet address pool is performed.

Sample config:

ManagementSubnet     = '10.10.10.0/24'



PASubnet             = '10.20.30.0/24'

PAPoolStart          = '10.20.30.100'

PAPoolEnd            = '10.20.30.200'



Gateways = @(

   @{ComputerName='gw01'; HostName='host01'; ManagementIP='10.10.10.131'; FrontEndIP='10.20.30.131'},

   @{ComputerName='gw02'; HostName='host02'; ManagementIP='10.10.10.132'; FrontEndIP='10.20.30.132'}

)

Expected behaviour:

  *   gw01 should use front-end IP address 10.20.30.131.
  *   gw02 should use front-end IP address 10.20.30.132.

Actual behaviour:

  *   Each gateway gets the first available IP address from the PA pool, starting from the end.

-

**Comment by MassimoPascucci** [comment]
@AnirbanPaul in the environment I'm currently working in, the customer wants full control of IP address allocation; so yes, we have a reason for this.

But, regardless of our reason for wanting to use it... if a feature is present and documented, it should _work_, shouldn't it?
Silently ignoring a configuration parameter and doing something else is definitely not the way to behave for a deployment script (or, really, any software).

If this is not a bug and was indeed done deliberately, the parameter should be removed from documentation and examples.
Otherwise, it should be taken into account and the script should behave accordingly.

**Comment by AnirbanPaul** [comment]
Yes, completely agree. The change has been done recently, and the docs have not been updated yet. We will get this changed ASAP.

**Comment by MassimoPascucci** [comment]
I strongly suggest to leave the option available.

Picking the highest available addresses from the pool is fine as a default; but if someone wants to put in the extra effort to explicitly select which addresses to use, the script should honour their wishes.

Also, this option is available for the management network; it makes sense to allow it for the front-end network too.

Last but not least, this has been possible for a while, even if few people used it; removing it is a breaking change, which should be avoided (unless there is a really strong reason for it).

**Comment by AnirbanPaul** [comment]
Thanks for your feedback. We will definitely take this into consideration.

**Comment by MassimoPascucci** [comment]
Also, please note that this is still an option for load balancers (parameter "PAIPAddress"); it makes sense to allow it for gateways, as it was indeed allowed until this change.

**Comment by MassimoPascucci** [comment]
Hello,

any news on this issue?

**Comment by AnirbanPaul** [comment]
Hi Massimo
This is still in our backlog. I do not have any publish dates for this work item right now.
